overlord/devicestate/devicemgr.go

```diff
@@ -148,31 +151,82 @@ func (m *DeviceManager) ensureOperational() error {
 		return nil
 	}
 
+	// conditions to trigger device registration
+	//
+	// * have a model assertion with a gadget (core and
+	//   device-like classic) in which case we need also to wait
+	//   for the gadget to have been installed though
+	// * classic with a model assertion with a non-default store specified
+	// * lazy classic case (might have a model with no gadget nor store
+	//   or no model): we wait to have some snaps installed or be
+	//   in the process to install some
+
+	var seeded bool
+	err = m.state.Get("seeded", &seeded)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+
 	if device.Brand == "" || device.Model == "" {
-		// cannot proceed until seeding has loaded the model
-		// assertion and set the brand and model, that is
-		// optional on classic
-		// TODO: later we can check if
-		// "seeded" was set and we still don't have a brand/model
-		// and use a fallback assertion
-		return nil
+		if !release.OnClassic || !seeded {
+			return nil
+		}
+		// we are on classic and seeded but there is no model:
+		// use a fallback model!
+		err = assertstate.Add(m.state, sysdb.GenericClassicModel())
+		if err != nil && !asserts.IsUnaccceptedUpdate(err) {
+			return fmt.Errorf(`cannot install "generic-classic" fallback model assertion: %v`, err)
+		}
+		device.Brand = "generic"
+		device.Model = "generic-classic"
+		if err := auth.SetDevice(m.state, device); err != nil {
+			return err
+		}
 	}
 
 	if m.changeInFlight("become-operational") {
 		return nil
 	}
 
-	// TODO: make presence of gadget optional on classic? that is
-	// sensible only for devices that the store can give directly
-	// serials to and when we will have a general fallback
-	gadgetInfo, err := snapstate.GadgetInfo(m.state)
-	if err == state.ErrNoState {
-		// no gadget installed yet, cannot proceed
-		return nil
-	}
-	if err != nil {
+	var storeID, gadget string
+	model, err := Model(m.state)
+	if err != nil && err != state.ErrNoState {
 		return err
 	}
+	if err == nil {
+		gadget = model.Gadget()
+		storeID = model.Store()
+	} else {
+		return fmt.Errorf("internal error: core device brand and model are set but there is no model assertion")
+	}
+
+	if gadget == "" && storeID == "" {
+		// classic: if we have no gadget and no non-default store
+		// wait to have snaps or snap installation
+
+		// TODO: implement snapstate.NumSnaps!
```

Comment:
Indeed.. or even HasSnaps, which may be cheaper than counting (e.g. may be implemented in terms of core).